lib/internal/modules/cjs/loader.js

```diff
@@ -357,6 +359,126 @@ function findLongestRegisteredExtension(filename) {
   return '.js';
 }
 
+function resolveBasePath(basePath, exts, isMain, trailingSlash, request) {
+  let filename;
+
+  const rc = stat(basePath);
+  if (!trailingSlash) {
+    if (rc === 0) {  // File.
+      if (!isMain) {
+        if (preserveSymlinks) {
+          filename = path.resolve(basePath);
+        } else {
+          filename = toRealPath(basePath);
+        }
+      } else if (preserveSymlinksMain) {
+        // For the main module, we use the preserveSymlinksMain flag instead
+        // mainly for backward compatibility, as the preserveSymlinks flag
+        // historically has not applied to the main module.  Most likely this
+        // was intended to keep .bin/ binaries working, as following those
+        // symlinks is usually required for the imports in the corresponding
+        // files to resolve; that said, in some use cases following symlinks
+        // causes bigger problems which is why the preserveSymlinksMain option
+        // is needed.
+        filename = path.resolve(basePath);
+      } else {
+        filename = toRealPath(basePath);
+      }
+    }
+
+    if (!filename) {
+      // Try it with each of the extensions
+      if (exts === undefined)
+        exts = Object.keys(Module._extensions);
+      filename = tryExtensions(basePath, exts, isMain);
+    }
+  }
+
+  if (!filename && rc === 1) {  // Directory.
+    // try it with each of the extensions at "index"
+    if (exts === undefined)
+      exts = Object.keys(Module._extensions);
+    filename = tryPackage(basePath, exts, isMain, request);
+  }
+
+  return filename;
+}
+
+// This only applies to requests of a specific form:
+// 1. @
+// 2. @/.*
+const SELF_PATTERN = /^@(\/.*)?$/;
+function trySelf(paths, exts, isMain, trailingSlash, request) {
+  if (!experimentalExports) {
+    return false;
+  }
+
+  const [, expansion] =
+    StringPrototype.match(request, SELF_PATTERN) || [];
+
+  if (!expansion) {
+    return false;
+  }
```

Comment:
Wouldn't it be faster and simpler in these 5 lines to do this (and drop `SELF_PATTERN`)? The regexp just seems unnecessarily clever to me.

```js
if (request !== "@" && !StringPrototype.startsWith(request, "@/")) {
	return false
}

const expansion = StringPrototype.slice(request, 1)
```

The `request !== "@"` is a simple constant comparison as V8 preallocates all 1-character strings, and the rest is basically what the regexp algorithm would have to do anyways, just without all the ceremony of building and executing the regexp itself.

Reply:
>  is a simple constant comparison as V8 preallocates all 1-character strings,

This is a really small nit and your suggestion is still solid: "all 1-character strings between 0 and 128"

Reply:
🤦‍♀️ Thanks for the correction!

Reply:
Good idea, I think that also matches the C++ implementation more closely. Updated.

Reply:
Good idea, I think that also matches the C++ implementation more closely. Updated.